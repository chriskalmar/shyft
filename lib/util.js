
import DataType from './datatype/DataType';
import Entity from './entity/Entity';


export const passOrThrow = (condition, messageFn) => {
  // providing the error message as a callback return massively improves code coverage reports.
  // the message function is only evaluated if the condition fails, which will show up correctly
  // in the coverage reports.
  if (typeof messageFn !== 'function') {
    throw new Error('passOrThrow() expects messageFn to be a function')
  }
  if (!condition) {
    throw new Error(messageFn())
  }
}


export const isDataType = (obj) => {
  return (obj instanceof DataType)
}


export const isEntity = (obj) => {
  return (obj instanceof Entity)
}


export const resolveFunctionMap = (functionOrMap) => {
  return typeof functionOrMap === 'function'
    ? functionOrMap()
    : functionOrMap
}


export const isMap = (map, nonEmpty) => {
  return map !== null &&
    typeof map === 'object' &&
    Array.isArray(map) === false &&
    (!nonEmpty || (nonEmpty && Object.keys(map).length > 0 ))
}


export const isFunction = (fn) => {
  return typeof fn === 'function'
}



export const isArray = (set, nonEmpty) => {
  return set !== null &&
    typeof map === 'array' &&
    Array.isArray(set) === true &&
    (!nonEmpty || (nonEmpty && set.length > 0 ))
}

