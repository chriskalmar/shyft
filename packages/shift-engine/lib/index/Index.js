
import {
  passOrThrow,
  isArray,
} from '../util';

import _ from 'lodash';

export const INDEX_UNIQUE = 'unique';
export const INDEX_GENERIC = 'generic';
export const indexTypes = [
  INDEX_UNIQUE,
  INDEX_GENERIC,
]


class Index {

  constructor (setup = {}) {

    const {
      type,
      attributes,
    } = setup

    passOrThrow(type, () => 'Missing index type')
    passOrThrow(
      indexTypes.indexOf(type) >= 0,
      () => `Unknown index type '${type}' used, try one of these: '${indexTypes.join(', ')}'`
    )

    passOrThrow(
      isArray(attributes, true),
      () => `Index definition of type '${type}' needs to have a list of attributes`
    )

    attributes.map(attribute => {
      passOrThrow(
        typeof attribute === 'string',
        () => `Index definition of type '${type}' needs to have a list of attribute names`
      )
    })

    passOrThrow(
      attributes.length === _.uniq(attributes).length,
      () => `Index definition of type '${type}' needs to have a list of unique attribute names`
    )


    this.type = type
    this.attributes = attributes

  }


  toString() {
    return this.type
  }

}


export default Index


export const isIndex = (obj) => {
  return (obj instanceof Index)
}


export const processEntityIndexes = (entity, indexes) => {

  passOrThrow(
    isArray(indexes),
    () => `Entity '${entity.name}' indexes definition needs to be an array of indexes`
  )


  const singleAttributeIndexes = []

  const entityAttributes = entity.getAttributes()

  indexes.map((index, idx) => {
    passOrThrow(
      isIndex(index),
      () => `Invalid index definition for entity '${entity.name}' at position '${idx}'`
    )


    index.attributes.map((attributeName) => {

      passOrThrow(
        entityAttributes[ attributeName ],
        () => `Cannot use attribute '${entity.name}.${attributeName}' in index as it does not exist`
      )


      if (index.type === INDEX_UNIQUE) {

        passOrThrow(
          entityAttributes[ attributeName ].required,
          () => `Cannot use attribute '${entity.name}.${attributeName}' in uniqueness index as it is nullable`
        )

        passOrThrow(
          !entityAttributes[ attributeName ].i18n,
          () => `Cannot use attribute '${entity.name}.${attributeName}' in uniqueness index as it is translatable`
        )

        if (index.attributes.length === 1) {
          entityAttributes[ attributeName ].isUnique = true
        }
      }

      if (index.attributes.length === 1) {
        singleAttributeIndexes.push(attributeName)
      }

    })
  })

  // add new index for single attributes with the 'index' flag set but not defined in an index object by the user
  Object.keys(entityAttributes).map(attributeName => {
    if (entityAttributes[attributeName].index) {
      if (!singleAttributeIndexes.includes(attributeName)) {
        indexes.push(new Index({
          type: INDEX_GENERIC,
          attributes: [attributeName]
        }))
      }
    }
  })

  return indexes
}
