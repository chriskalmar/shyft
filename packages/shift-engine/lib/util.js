import casual from 'casual';
import _ from 'lodash';



export const deleteUndefinedProps = (obj) =>
  Object.keys(obj).forEach(key => obj[key] === undefined && delete obj[key])



export const passOrThrow = (condition, messageFn) => {
  // providing the error message as a callback return massively improves code coverage reports.
  // the message function is only evaluated if the condition fails, which will show up correctly
  // in the coverage reports.
  if (typeof messageFn !== 'function') {
    throw new Error('passOrThrow() expects messageFn to be a function')
  }
  if (!condition) {
    throw new Error(messageFn())
  }
}


export const resolveFunctionMap = (functionOrMap) => {
  return typeof functionOrMap === 'function'
    ? functionOrMap()
    : { ...functionOrMap }
}


export const isMap = (map, nonEmpty) => {
  return map !== null &&
    typeof map === 'object' &&
    Array.isArray(map) === false &&
    (!nonEmpty || (nonEmpty && Object.keys(map).length > 0 ))
}


export const isFunction = (fn) => {
  return typeof fn === 'function'
}



export const isArray = (set, nonEmpty) => {
  return set !== null &&
    Array.isArray(set) === true &&
    (!nonEmpty || (nonEmpty && set.length > 0 ))
}


export const isString = (str) => (typeof str === 'string')


export const mergeMaps = (first, second) => {
  if (!isMap(first) || !isMap(second)) {
    throw new Error('mergeMaps() expects 2 maps for a merge to work')
  }

  return { ...first, ...second }
}



export const mapOverProperties = (object, iteratee) => {

  passOrThrow(
    isMap(object),
    () => 'Provided object is not a map'
  )

  passOrThrow(
    isFunction(iteratee),
    () => 'Provided iteratee is not a function'
  )


  const keys = Object.keys(object);

  keys.forEach((key) => {
    iteratee(object[ key ], key)
  })
}



export const sortDataByKeys = (keys, data, keyProperty='id') => {
  const map = {}
  const result = []
  const order = {}

  if (!keys || ( isArray(keys) && keys.length === 0)) {
    return []
  }

  if (!data || ( isArray(data) && data.length === 0)) {
    return keys.map(() => null)
  }


  for (let d = 0; d < data.length; d++) {
    const row = data[ d ]
    const id = row[ keyProperty ]
    const found = map[id]

    if (!found) {
      map[id] = row
    }
    else if (found instanceof Array) {
      found.push(row)
    }
    else {
      map[id] = [ found, row ];
    }
  }


  for (let k = 0; k < keys.length; k++) {
    const key = keys[ k ]
    const found = map[ key ];

    if (typeof found === 'undefined') {
      result.push(null)
    }
    else if (found instanceof Array) {
      let idx = order[ key ];

      if (typeof idx === 'undefined') {
        idx = 0
      }
      else {
        idx++
      }

      order[ key ] = idx;

      result.push(found[ idx ]);
    }
    else {
      result.push(found);
    }
  }

  return result;
};



export const processCursor = (entity, cursor, orderBy, reverse) => {

  const $LT = reverse ? '$gt' : '$lt'
  const $GT = reverse ? '$lt' : '$gt'
  const $LTE = reverse ? '$gte' : '$lte'
  const $GTE = reverse ? '$lte' : '$gte'

  const where = {}

  if (cursor) {
    passOrThrow(
      isArray(cursor[ entity.name ]),
      () => 'Incompatible cursor for this entity'
    )

    passOrThrow(
      isArray(orderBy),
      () => 'orderBy needs to be an array of order definitions'
    )

    const orderList = []

    const orderMap = {}
    orderBy.map(({attribute, direction}) => {

      passOrThrow(
        attribute && direction,
        () => 'orderBy needs to be an array of attributes and respective sort order'
      )

      orderMap[ attribute ] = direction
    })

    const primaryAttribute = entity.getPrimaryAttribute()
    const attributes = entity.getAttributes()

    let foundUniqueAttribute = false

    cursor[ entity.name ].map(filter => {
      if (filter.length !== 2) {
        throw new Error('Cursor malformed')
      }

      const attributeName = filter[0]

      passOrThrow(
        attributes[ attributeName ],
        () => `Unknown attribute '${attributeName}' used in cursor`
      )

      if (attributeName !== primaryAttribute.name) {
        passOrThrow(
          orderMap[ attributeName ],
          () => `Cursor works only on sorted attributes (check: '${attributeName}')`
        )
      }

      const attribute = attributes[ attributeName ]

      // limit where clause to the first attribute which is defined as unique
      if (!foundUniqueAttribute) {
        orderList.push(attributeName)
      }

      if (attribute.isUnique) {
        foundUniqueAttribute = true
      }

    })

    passOrThrow(
      foundUniqueAttribute,
      () => 'Cursor needs to have at least one attribute defined as unique'
    )


    // if more than 2 attributes are used for the cursor take only the first and the last (primary key)
    if (orderList.length > 2) {
      orderList.splice(1, (orderList.length - 2) )
    }


    // simple filter for single attributes
    if (orderList.length === 1) {

      const attributeName = orderList[0]
      const value = cursor[ entity.name ][0][1]

      if (orderMap[ attributeName ] === 'DESC') {
        where[ attributeName ] = {
          [ $LT ]: value
        }
      }
      else {
        where[ attributeName ] = {
          [ $GT ]: value
        }
      }

    }
    else {
      where.$not = {}

      cursor[ entity.name ].map(filter => {
        const attributeName = filter[0]
        const value = filter[1]

        // ignore attributes that obsolete due to prior unique attribute
        if (orderList.indexOf(attributeName) === -1) {
          return
        }

        const attribute = attributes[ attributeName ]

        if (attribute.isUnique) {
          if (orderMap[ attributeName ] === 'DESC') {
            where.$not[ attributeName ] = {
              [ $GTE ]: value
            }
          }
          else {
            where.$not[ attributeName ] = {
              [ $LTE ]: value
            }
          }
        }
        else {
          where.$not[ attributeName ] = value

          if (orderMap[ attributeName ] === 'DESC') {
            where[ attributeName ] = {
              [ $LTE ]: value
            }
          }
          else {
            where[ attributeName ] = {
              [ $GTE ]: value
            }
          }
        }

      })
    }
  }

  return where
}


export const processCursors = (entity, args) => {

  const {
    after,
    before,
  } = args;

  const where = {
    $and: [
      processCursor(entity, after, args.orderBy),
      processCursor(entity, before, args.orderBy, true),
    ]
  }

  return where
}



export const reverseString = (str) => {
  return str
    .split('')
    .reverse()
    .join('');
}



export const randomJson = () => {
  const ret = {}
  const count = casual.integer(1, 5)

  _.times(count, () => {
    const key = _.camelCase(casual.words(2))
    const value = Math.random() > 0.5
      ? casual.title
      : casual.integer()

    ret[key] = value
  })

  return ret
}
