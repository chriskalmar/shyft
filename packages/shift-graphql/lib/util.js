
import { constants } from 'shift-engine';
import _ from 'lodash';


// generate a type name from a model
export function generateTypeName(entityModel) {

  if (!entityModel || !_.isObject(entityModel)) {
    throw new Error('generateTypeName() entityModel needs to be defined')
  }

  if (!entityModel.name || !entityModel.domain) {
    throw new Error('generateTypeName() entityModel needs a name and a domain')
  }

  let name = entityModel.domain

  // if it's not the local provider use it for the node name (unique naming)
  if (constants.localProviderName !== entityModel.provider) {
    name = `${entityModel.provider}_${name}`
  }

  // add entity name
  name += `_${entityModel.name}`

  return _.camelCase(name)
}


export function upperCaseFirst(str) {
  return str.charAt(0).toUpperCase() + str.slice(1);
}



export function generateTypeNameUpperCase(entityModel) {
  return upperCaseFirst(
    generateTypeName(entityModel)
  )
}


export default {
  generateTypeName,
  upperCaseFirst,
  generateTypeNameUpperCase,
}
