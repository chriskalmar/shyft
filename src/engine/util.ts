import * as casual from 'casual';
import * as _ from 'lodash';
import { viewAttributePropertiesWhitelist } from './constants';

type StringFunction = () => string;

export const deleteUndefinedProps = (obj) =>
  Object.keys(obj).forEach((key) => obj[key] === undefined && delete obj[key]);

export const passOrThrow = (
  condition: any,
  messageFn: string | StringFunction,
): void => {
  // providing the error message as a callback return massively improves code coverage reports.
  // the message function is only evaluated if the condition fails, which will show up correctly
  // in the coverage reports.
  if (typeof messageFn !== 'function') {
    throw new Error('passOrThrow() expects messageFn to be a function');
  }
  if (!condition) {
    throw new Error(messageFn());
  }
};

export const resolveFunctionMap = (functionOrMap) => {
  return typeof functionOrMap === 'function'
    ? functionOrMap()
    : { ...functionOrMap };
};

export const isMap = (map: object, nonEmpty?: boolean): boolean => {
  return (
    map !== null &&
    typeof map === 'object' &&
    Array.isArray(map) === false &&
    (!nonEmpty || (nonEmpty && Object.keys(map).length > 0))
  );
};

export const isFunction = (fn: any): boolean => {
  return typeof fn === 'function';
};

export const isArray = (set: any[], nonEmpty?: boolean): boolean => {
  return (
    set !== null &&
    Array.isArray(set) === true &&
    (!nonEmpty || (nonEmpty && set.length > 0))
  );
};

export const isString = (str: any): boolean => typeof str === 'string';

export const mergeMaps = (first, second) => {
  if (!isMap(first) || !isMap(second)) {
    throw new Error('mergeMaps() expects 2 maps for a merge to work');
  }

  return { ...first, ...second };
};

export const mapOverProperties = (
  object: object,
  iteratee: (val: any, key: any) => any,
) => {
  passOrThrow(isMap(object), () => 'Provided object is not a map');

  passOrThrow(
    isFunction(iteratee),
    () => 'Provided iteratee is not a function',
  );

  const keys = Object.keys(object);

  keys.forEach((key) => {
    iteratee(object[key], key);
  });
};

export const sortDataByKeys = (
  keys: string[],
  data: object[],
  keyProperty = 'id',
) => {
  const map = {};
  const result = [];
  const order = {};

  if (!keys || (isArray(keys) && keys.length === 0)) {
    return [];
  }

  if (!data || (isArray(data) && data.length === 0)) {
    return keys.map(() => null);
  }

  for (let d = 0; d < data.length; d++) {
    const row = data[d];
    const id = row[keyProperty];
    const found = map[id];

    if (!found) {
      map[id] = row;
    } else if (found instanceof Array) {
      found.push(row);
    } else {
      map[id] = [found, row];
    }
  }

  for (let k = 0; k < keys.length; k++) {
    const key = keys[k];
    const found = map[key];

    if (typeof found === 'undefined') {
      result.push(null);
    } else if (found instanceof Array) {
      let idx = order[key];

      if (typeof idx === 'undefined') {
        idx = 0;
      } else {
        idx++;
      }

      order[key] = idx;

      result.push(found[idx]);
    } else {
      result.push(found);
    }
  }

  return result;
};

export const reverseString = (str) => {
  return str.split('').reverse().join('');
};

export const randomJson = () => {
  const ret = {};
  const count = casual.integer(1, 5);

  _.times(count, () => {
    const key = _.camelCase(casual.words(2));
    const value = Math.random() > 0.5 ? casual.title : casual.integer();

    ret[key] = value;
  });

  return ret;
};

export const asyncForEach = async (
  array: any[],
  callback: (val, idx, arr) => any,
) => {
  for (let index = 0; index < array.length; index++) {
    await callback(array[index], index, array);
  }
};

export const isDefined = (val: any): boolean => typeof val !== 'undefined';

export const convertEntityToViewAttribute = (attribute) => {
  if (!_.isObject(attribute)) {
    throw new Error(
      'convertEntityToViewAttribute() expects an attribute as input',
    );
  }

  const newAttribute = {};

  viewAttributePropertiesWhitelist.map((prop) => {
    newAttribute[prop] = attribute[prop];
  });

  return newAttribute;
};

export const convertEntityToViewAttributesMap = (attributesMap) => {
  const newAttributesMap = {};

  if (!_.isObject(attributesMap)) {
    throw new Error(
      'convertEntityToViewAttributesMap() expects an attributes map as input',
    );
  }

  for (const [name, attribute] of Object.entries(attributesMap)) {
    newAttributesMap[name] = convertEntityToViewAttribute(attribute);
  }

  return newAttributesMap;
};
