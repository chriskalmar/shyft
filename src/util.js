import crypto from 'crypto';

// postgres limitation
const SYSTEM_NAME_MAX_LENGTH = 63;

export const generateIndexName = (entityName, attributes, suffix = '_idx') => {
  let ret = `${entityName}_${suffix}_`;
  const attributesChain = (attributes || []).join('_');
  const attributesSortedChain = (attributes || []).sort().join('_');

  if (ret.length + attributesChain.length > SYSTEM_NAME_MAX_LENGTH) {
    const hashLength = 10;
    const hash = crypto
      .createHash('sha256')
      .update(attributesSortedChain)
      .digest('hex')
      .substr(0, hashLength);

    const shortendAttributesChain = attributesChain.substr(
      0,
      SYSTEM_NAME_MAX_LENGTH - hashLength - ret.length - 1,
    );

    ret = `${entityName}_${shortendAttributesChain}_${hash}_${suffix}`;
  } else {
    ret = `${entityName}_${attributesChain}_${suffix}`;
  }

  return ret;
};

export const invertDirection = direction => {
  return direction === 'DESC' ? 'ASC' : 'DESC';
};

export const getLimit = (first, last) => {
  if (first >= 0) {
    return first;
  } else if (last >= 0) {
    return last;
  }

  return 10;
};

export const asyncForEach = async (array, callback) => {
  for (let index = 0; index < array.length; index++) {
    await callback(array[index], index, array);
  }
};

export const quote = item => {
  if (!item) {
    throw new Error('quote() requires an input');
  }

  // special care for json pointers
  if (item.includes('->')) {
    const arr = item.split('->');
    arr[0] = `"${arr[0]}"`;
    return arr.join('->');
  }

  return `"${item}"`;
};
