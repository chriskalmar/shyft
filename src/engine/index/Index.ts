import { uniq } from 'lodash';
import { passOrThrow, isArray } from '../util';

import { Entity } from '../entity/Entity';

export const INDEX_UNIQUE = 'unique';
export const INDEX_GENERIC = 'generic';
export const indexTypes = [INDEX_UNIQUE, INDEX_GENERIC];

export type IndexSetup = {
  type?: string;
  attributes?: string[];
};

export class Index {
  type: string;
  attributes: string[];

  constructor(setup: IndexSetup = {} as IndexSetup) {
    const { type, attributes } = setup;

    passOrThrow(type, () => 'Missing index type');
    passOrThrow(
      indexTypes.indexOf(type) >= 0,
      () =>
        `Unknown index type '${type}' used, try one of these: '${indexTypes.join(
          ', ',
        )}'`,
    );

    passOrThrow(
      isArray(attributes, true),
      () =>
        `Index definition of type '${type}' needs to have a list of attributes`,
    );

    attributes.map((attribute) => {
      passOrThrow(
        typeof attribute === 'string',
        () =>
          `Index definition of type '${type}' needs to have a list of attribute names`,
      );
    });

    passOrThrow(
      attributes.length === uniq(attributes).length,
      () =>
        `Index definition of type '${type}' needs to have a list of unique attribute names`,
    );

    this.type = type;
    this.attributes = attributes;
  }

  toString() {
    return this.type;
  }
}

export const isIndex = (obj: unknown): obj is Index => {
  return obj instanceof Index;
};

export const processEntityIndexes = (entity: Entity, indexes: Index[]) => {
  passOrThrow(
    isArray(indexes),
    () =>
      `Entity '${entity.name}' indexes definition needs to be an array of indexes`,
  );

  const singleAttributeIndexes = [];

  const entityAttributes = entity.getAttributes();

  indexes.map((index, idx) => {
    passOrThrow(
      isIndex(index),
      () =>
        `Invalid index definition for entity '${entity.name}' at position '${idx}'`,
    );

    index.attributes.map((attributeName) => {
      passOrThrow(
        entityAttributes[attributeName],
        () =>
          `Cannot use attribute '${entity.name}.${attributeName}' in index as it does not exist`,
      );

      if (index.type === INDEX_UNIQUE) {
        passOrThrow(
          entityAttributes[attributeName].required,
          () =>
            `Cannot use attribute '${entity.name}.${attributeName}' in uniqueness index as it is nullable`,
        );

        passOrThrow(
          !entityAttributes[attributeName].i18n,
          () =>
            `Cannot use attribute '${entity.name}.${attributeName}' in uniqueness index as it is translatable`,
        );

        if (index.attributes.length === 1) {
          entityAttributes[attributeName].unique = true;
        }
      }

      if (index.attributes.length === 1) {
        singleAttributeIndexes.push(attributeName);
      }
    });
  });

  // add new index for single attributes with the 'index' flag set but not defined in an index object by the user
  Object.keys(entityAttributes).map((attributeName) => {
    if (entityAttributes[attributeName].index) {
      if (!singleAttributeIndexes.includes(attributeName)) {
        indexes.push(
          new Index({
            type: INDEX_GENERIC,
            attributes: [attributeName],
          }),
        );
      }
    }
  });

  return indexes;
};
