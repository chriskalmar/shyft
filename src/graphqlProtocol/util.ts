import * as _ from 'lodash';
import * as pluralize from 'pluralize';
import { RELAY_TYPE_PROMOTER_FIELD } from './protocolGraphqlConstants';
import { CustomError } from '../engine/CustomError';
import { isDefined } from '../engine/util';
import { getRegisteredEntityAttribute } from './registry';

export function generateTypeName(name) {
  return _.camelCase(name);
}

export function pascalCase(str) {
  return str.charAt(0).toUpperCase() + str.slice(1);
}

export function plural(str) {
  return pluralize.plural(str);
}

export function generateTypeNamePascalCase(name) {
  return pascalCase(generateTypeName(name));
}

export function generateTypeNamePlural(name) {
  return pluralize.plural(generateTypeName(name));
}

export function generateTypeNamePluralPascalCase(name) {
  return pascalCase(pluralize.plural(generateTypeName(name)));
}

export function generateTypeNameUpperCase(name) {
  return _.snakeCase(name).toUpperCase();
}

export const toBase64 = (value) =>
  new Buffer(value.toString()).toString('base64');
export const fromBase64 = (value) =>
  new Buffer(value.toString(), 'base64').toString();

export const serializeCursor = (cursor) => toBase64(JSON.stringify(cursor));
export const deserializeCursor = (cursor) => {
  try {
    return JSON.parse(fromBase64(cursor));
  } catch (e) {
    // TODO: check
    // should throw error, not return
    // https://github.com/graphql/graphql-js/issues/910
    return new Error('Invalid cursor provided');
  }
};

export const addRelayTypePromoterToInstance = (typeName, instance) => {
  if (!instance) {
    return instance;
  }

  instance[RELAY_TYPE_PROMOTER_FIELD] = typeName;
  return instance;
};

export const addRelayTypePromoterToList = (typeName, list) => {
  return list.map((instance) => {
    return addRelayTypePromoterToInstance(typeName, instance);
  });
};

export const addRelayTypePromoterToInstanceFn = (typeName) => {
  return (instance) => {
    return addRelayTypePromoterToInstance(typeName, instance);
  };
};

export const addRelayTypePromoterToListFn = (typeName) => {
  return (list) => {
    return addRelayTypePromoterToList(typeName, list);
  };
};

export const translateInstance = (
  entity,
  instance,
  { i18nLanguage, i18nDefaultLanguage, i18nLanguageDisableFallback },
) => {
  if (
    !instance ||
    !i18nLanguage ||
    i18nLanguage === i18nDefaultLanguage ||
    !entity.getI18nAttributeNames ||
    !entity.getI18nAttributeNames()
  ) {
    return instance;
  }

  const attributes = entity.getAttributes();
  const i18nAttributeNames = entity.getI18nAttributeNames();

  i18nAttributeNames.map((attributeName) => {
    const { required } = attributes[attributeName];
    const {
      fieldName: gqlFieldName,
      fieldNameI18n: gqlFieldNameI18n,
    } = getRegisteredEntityAttribute(
      entity.name,
      attributes[attributeName].name,
    );

    const translation = instance[gqlFieldNameI18n]
      ? instance[gqlFieldNameI18n][i18nLanguage]
      : undefined;

    if (i18nLanguageDisableFallback) {
      instance[gqlFieldName] =
        !isDefined(translation) && required
          ? new CustomError(
              `Translation for '${gqlFieldName}' not found`,
              'TranslationNotFoundError',
            )
          : translation;
    } else if (isDefined(translation)) {
      instance[gqlFieldName] = translation;
    }
  });

  return instance;
};

export const translateList = (entity, list, context) => {
  return list.map((instance) => {
    return translateInstance(entity, instance, context);
  });
};

export const translateInstanceFn = (entity, context) => {
  return (instance) => {
    return translateInstance(entity, instance, context);
  };
};

export const translateListFn = (entity, context) => {
  return (list) => {
    return translateList(entity, list, context);
  };
};

// emulate gql tag for prettier
export const gql = String.raw;
