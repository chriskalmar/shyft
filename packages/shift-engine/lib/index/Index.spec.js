
import { assert } from 'chai';
import Index, {
  isIndex,
  INDEX_UNIQUE,
} from './Index';
import { passOrThrow } from '../util';



describe('Index', () => {

  it('should have a type', () => {

    function fn() {
      new Index() // eslint-disable-line no-new
    }

    assert.throws(fn, /Missing index type/);

  })


  it('should have a valid type', () => {

    function fn() {
      new Index({ // eslint-disable-line no-new
        type: 'something'
      })
    }

    assert.throws(fn, /Unknown index type/);

  })


  it('should have a list of attributes', () => {

    function fn1() {
      new Index({ // eslint-disable-line no-new
        type: INDEX_UNIQUE
      })
    }

    assert.throws(fn1, /needs to have a list of attributes/);

    function fn2() {
      new Index({ // eslint-disable-line no-new
        type: INDEX_UNIQUE,
        attributes: [ ]
      })
    }

    assert.throws(fn2, /needs to have a list of attributes/);

  })



  it('should accept attribute names only', () => {

    function fn1() {
      new Index({ // eslint-disable-line no-new
        type: INDEX_UNIQUE,
        attributes: [ null ]
      })
    }

    assert.throws(fn1, /needs to have a list of attribute names/);

    function fn2() {
      new Index({ // eslint-disable-line no-new
        type: INDEX_UNIQUE,
        attributes: [ 123 ]
      })
    }

    assert.throws(fn2, /needs to have a list of attribute names/);

  })


  it('should accept unique attribute names only', () => {

    function fn() {
      new Index({ // eslint-disable-line no-new
        type: INDEX_UNIQUE,
        attributes: [ 'a', 'b', 'a' ]
      })
    }

    assert.throws(fn, /needs to have a unique list of attribute names/);

  })


  it('should accept a correct definition', () => {

    const index1 = new Index({
      type: INDEX_UNIQUE,
      attributes: [ 'a' ]
    })

    const index2 = new Index({
      type: INDEX_UNIQUE,
      attributes: [ 'a', 'b', 'c' ]
    })

    assert.deepEqual(index1.attributes, [ 'a' ]);
    assert.deepEqual(index2.attributes, [ 'a', 'b', 'c' ]);

  })



  describe('isIndex', () => {


    it('should recognize objects of type Index', () => {

      const index = new Index({
        type: INDEX_UNIQUE,
        attributes: [ 'a' ]
      })

      function fn() {
        passOrThrow(
          isIndex(index),
          () => 'This error will never happen'
        )
      }

      assert.doesNotThrow(fn)

    })


    it('should recognize non-Index objects', () => {

      function fn() {
        passOrThrow(
          isIndex({}) ||
          isIndex(function test() {}) ||
          isIndex(assert),
          () => 'Not an Index object'
        )
      }


      assert.throws(fn, /Not an Index object/);

    })

  })


})
