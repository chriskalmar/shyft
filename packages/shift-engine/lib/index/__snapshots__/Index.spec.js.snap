// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Index isIndex should recognize non-Index objects 1`] = `"Not an Index object"`;

exports[`Index processEntityIndexes should throw if attribute used in index does not exist 1`] = `"Cannot use attribute 'SomeEntityName.notHere' in index as it does not exist"`;

exports[`Index processEntityIndexes should throw if provided with an invalid index 1`] = `"Invalid index definition for entity 'SomeEntityName' at position '0'"`;

exports[`Index processEntityIndexes should throw if provided with an invalid list of indexes 1`] = `"Entity 'SomeEntityName' indexes definition needs to be an array of indexes"`;

exports[`Index should accept attribute names only 1`] = `"Index definition of type 'unique' needs to have a list of attribute names"`;

exports[`Index should accept attribute names only 2`] = `"Index definition of type 'unique' needs to have a list of attribute names"`;

exports[`Index should accept unique attribute names only 1`] = `"Index definition of type 'unique' needs to have a list of unique attribute names"`;

exports[`Index should have a list of attributes 1`] = `"Index definition of type 'unique' needs to have a list of attributes"`;

exports[`Index should have a list of attributes 2`] = `"Index definition of type 'unique' needs to have a list of attributes"`;

exports[`Index should have a type 1`] = `"Missing index type"`;

exports[`Index should have a valid type 1`] = `"Unknown index type 'something' used, try one of these: 'unique, generic'"`;
